src/main/kotlin/org/elm/ide/components/ElmFormatOnFileSaveComponent.kt

```diff
@@ -0,0 +1,58 @@
+package org.elm.ide.components
+
+import com.intellij.AppTopics
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.components.ProjectComponent
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.openapi.fileEditor.FileDocumentManagerListener
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.project.ProjectManager
+import com.intellij.openapi.roots.ProjectRootManager
+import com.intellij.openapi.vfs.VirtualFile
+import org.elm.lang.core.psi.isElmFile
+import org.elm.workspace.ElmFormatCLI
+import org.elm.workspace.elmSettings
+import org.elm.workspace.elmToolchain
+
+class ElmFormatOnFileSaveComponent : ProjectComponent {
+
+
+    override fun initComponent() {
+        val application = ApplicationManager.getApplication()
+        val bus = application.messageBus
+
+        bus.connect().subscribe(
+                AppTopics.FILE_DOCUMENT_SYNC,
+                object : FileDocumentManagerListener {
+                    override fun beforeDocumentSaving(document: Document) {
+
+                        val vFile = FileDocumentManager.getInstance().getFile(document) ?: return
+                        if (!vFile.isElmFile)
+                            return
+
+                        val project = projectForFile(vFile) ?: return
+
+                        if (project.elmSettings.toolchain?.activateElmFormatOnSaveHook != true) return
+
+                        val elmVersion = ElmFormatCLI.getElmVersion(project, vFile) ?: return
+
+                        val elmFormat = project.elmToolchain?.elmFormat ?: return
+                        elmFormat.formatDocumentAndSetText(project, document, elmVersion)
```

Comment:
Is it ok to block in this `beforeDocumentSaving` callback? Because we might take as long as 2 seconds to invoke elm-format. Is this was the Scala plugin does?

Reply:
The scalafmt-plugin does it exactly like that.
https://github.com/scalameta/scalafmt/blob/master/scalafmt-intellij/src/main/scala/org/scalafmt/intellij/OnFileSaveComponent.scala#L27

I don't think the scala-plugin by jetbrains does format-on-save-with-scalafmt.